enhancements/cluster-logging/multi-container-structured-logging.md

```diff
@@ -22,120 +22,135 @@ see-also:
+This proposal extends `output.elasticsearch` to allow the `structuredType` for different
+containers to be specified via _annotations_.
 
 ## Motivation
 
-Initial motivation comes from [forwarding JSON to Elasticsearch](forwarding-json-structured-logs.md).
-The popularity of JSON logging means that an application container be deployed in a pod with a sidecar logging an incompatible JSON log format.
-Adding `elasticsearch.structuredTypePattern` allows JSON logs from different containers to be sent to separate indices.
+- Sidecars are a common pattern in kubernetes and openshift clusters, which means multiple containers in a Pod.
+- Many popular sidecars (for example ISTIO) JSON logging.
+- The JSON log formats of sidecars and application containers may not be compatible, and must be separated to avoid index problems with Elasticsearch.
+- K8s annotations are used to [https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/](attach arbitrary non-identifying metadata to objects)
 
-This feature can be used in any situation where per-log output parameters are derived from log record fields.
-For example `cloudwatch.groupPattern`, `loki.tenantPattern`, `kafka.topicPattern` and so on.
 
 ### Goals
 
-Implement `elasticsearch.structuredTypePattern` for the Elasticsearch output.
+When JSON log records are forwarded Elasticsearch, direct messages with different JSON schema to different Elasticsearch indices.
 
 ### Non-Goals
 
-Other outputs may re-use this pattern, but are not required to complete this proposal.
+No other outputs will use this pattern initially, but may do in future if appropriate.
 
 ## Proposal
+### Example
 
-Add field `output.elasticsearch.structuredTypePattern` of type string.
-Compute the index name for each log record by applying the `structuredTypePattern` expression to the log record.
+An example will help.
 
-### User Stories
+``` yaml
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+metadata:
+  name: "instance"
+spec:
+  pipelines:
+  - name: JSONToElasticsearch
+    inputRefs: [ application ]
+	outputRefs: [ default ]
+	parse: json
+
+  outputDefaults:
+    - elasticsearch:
+	    structuredTypeAnnotation: foo.example.com/
+```
 
-#### Send JSON logs from pods labeled app=myApp to separate indices based on container name
+The forwarder configuration above parses and forwards JSON logs to the default elasticsearch
+instance. It uses the annotation prefix "foo.example.com/" to determine the
+structured type of each record. Here's an example deployment that shows how the
+annotations are used:
 
 ``` yaml
-outputs:
-  - name: myApp
-    type: elasticsearch
-	structuredTypePattern: '{.kubernetes.label.app.myApp}-{.kubernetes.container_name}'
+apiversion: apps/v1
+kind: Deployment
+metadata:
+  name: application-foo
+  annotations:
+	foo.example.com/application-foo: schema-v1
+	foo.example.com/sidecar-foo: schema-v2
+	foo.example.com/istio-sidecar: istio-schema-v1
```

Comment:
I'm interested in the implementation of this:

* What is an example of a concrete index?
* What does the rollover managment look like? 
* Is it falling under `app` and subject to the same criteria?  
* How is it aliased?

Reply:
This would be the same way we currently handle `structuredTypeKey`; except that instead of looking up the key as a label, we look up "prefix+container-name" as an annotation.